pkg/policy/l4.go

```diff
@@ -1084,13 +1111,280 @@ func addL4Filter(policyCtx PolicyContext,
 		}
 	}
 
-	resMap[key] = existingFilter
+	resMap.Upsert(p.Port, uint16(p.EndPort), string(proto), existingFilter)
 	return nil
 }
 
-// L4PolicyMap is a list of L4 filters indexable by protocol/port
-// key format: "port/proto"
-type L4PolicyMap map[string]*L4Filter
+// L4PolicyMap is a list of L4 filters indexable by port/endport/protocol
+type L4PolicyMap interface {
+	Upsert(port string, endPort uint16, protocol string, l4 *L4Filter)
+	Delete(port string, endPort uint16, protocol string)
+	ExactLookup(port string, endPort uint16, protocol string) *L4Filter
+	LongestPrefixMatch(port string, protocol string) *L4Filter
+	Detach(selectorCache *SelectorCache)
+	IngressCoversContext(ctx *SearchContext) api.Decision
+	EgressCoversContext(ctx *SearchContext) api.Decision
+	ForEach(func(l4 *L4Filter) bool)
+	Equals(t *testing.T, bMap L4PolicyMap) bool
+	Diff(t *testing.T, expectedMap L4PolicyMap) string
+	Len() int
+}
+
+// NewL4PolicyMap creates an new L4PolicMap.
+func NewL4PolicyMap() L4PolicyMap {
+	return &l4PolicyMap{
+		namedPortMap: make(map[string]*L4Filter),
+		rangePortMap: bitlpm.NewUintTrie[uint32, map[portProtoKey]*L4Filter](),
+	}
+}
+
+// InitL4PolicyMap creates an new L4PolicMap, with an initial
+// set of values. The initMap argument does not support ranges.
+func InitL4PolicyMap(initMap map[string]*L4Filter) L4PolicyMap {
+	l4M := &l4PolicyMap{
+		namedPortMap: make(map[string]*L4Filter),
+		rangePortMap: bitlpm.NewUintTrie[uint32, map[portProtoKey]*L4Filter](),
+	}
+	for k, v := range initMap {
+		portProtoSlice := strings.Split(k, "/")
+		if len(portProtoSlice) < 2 {
+			continue
+		}
+		l4M.Upsert(portProtoSlice[0], 0, portProtoSlice[1], v)
+	}
+	return l4M
+}
+
+type portProtoKey struct {
+	port, endPort uint16
+	proto         uint8
+}
+
+// l4PolicyMap is the implementation of L4PolicyMap
+type l4PolicyMap struct {
+	namedPortMap map[string]*L4Filter
+	// rangePortMap has to keep a map of L4Filters rather than
+	// a single L4Filter reference so that the l4PolicyMap does
+	// not merge L4Filter that are not the same port range, but
+	// share an overlapping range in the trie.
+	rangePortMap *bitlpm.UintTrie[uint32, map[portProtoKey]*L4Filter]
+	// rangeMapLen must be tracked separately from
+	// rangePortMap as L4Filters are split up when
+	// the port range is long enough. This does not
+	// represent additional entries only the vagaries
+	// of indexing by range.
```

Comment:
nit: I'm a bit confused what the "additional entries" are or what "only the vagaries of indexing by range" means.

Is the following a true statement?
> "rangeMapLen counts the number of unique L4Filters in the `rangePortMap`"

Reply:
Yes. "The vagaries of indexing by range" is an opaque reference to the idea that we can't have a single entry for one range because bitwise math does not allow it, so we need to split the range up into range lengths (divisors) of powers of two. I will try to document this better.